{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import time as t\n",
    "from collections import Counter\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [],
   "source": [
    "# implementation of avl tree\n",
    "\n",
    "class Node(object):\n",
    "    def __init__(self, data, val, res_ind, sp, var):\n",
    "        self.variant = var\n",
    "        self.species = sp\n",
    "        self.data = data\n",
    "        self.value = val\n",
    "        self.res_ind = res_ind\n",
    "        self.parent = None\n",
    "        self.height = 0\n",
    "        self.left_child = None\n",
    "        self.right_child = None\n",
    "\n",
    "\n",
    "class AVLTree(object):\n",
    "\n",
    "    def __init__(self):\n",
    "        self.root = None\n",
    "            \n",
    "            \n",
    "    def insert(self, data, val, res_ind, sp, var):\n",
    "        if self.root is None:\n",
    "            self.root = Node(data, val, res_ind, sp, var)\n",
    "        else:\n",
    "            new_node = self._insert(self.root, data, val, res_ind, sp, var)\n",
    "            self._walk_up(new_node)\n",
    "\n",
    "    def _insert(self, node, data, val, res_ind, sp, var):\n",
    "        if data <= node.data:\n",
    "            if not node.left_child:\n",
    "                node.left_child = Node(data, val, res_ind, sp, var)\n",
    "                node.left_child.parent = node\n",
    "                return node.left_child\n",
    "            return self._insert(node.left_child, data, val, res_ind, sp, var)\n",
    "        else:\n",
    "            if not node.right_child:\n",
    "                node.right_child = Node(data, val, res_ind, sp, var)\n",
    "                node.right_child.parent = node\n",
    "                return node.right_child\n",
    "            return self._insert(node.right_child, data, val, res_ind, sp, var)\n",
    "\n",
    "    def remove(self, data):\n",
    "        if not self.root:\n",
    "            raise ValueError('Tree is empty.')\n",
    "        self._delete_value(data)\n",
    "\n",
    "    def _delete_value(self, data):\n",
    "        node = self.find(data, self.root)\n",
    "        if node is False:\n",
    "            raise ValueError('No node with value {}'.format(data))\n",
    "        parent_of_deleted_node = self._delete_node(node)\n",
    "        # if parent is None we know that we just deleted the root node,\n",
    "        # but if root node had a child, that child is now the root node!\n",
    "        if not parent_of_deleted_node and self.root:\n",
    "            parent_of_deleted_node = self.root\n",
    "        self._walk_up(parent_of_deleted_node)\n",
    "\n",
    "    def find(self, value, node):\n",
    "        if node is None:\n",
    "            return False\n",
    "        if node.data > value:\n",
    "            return self.find(value, node.left_child)\n",
    "        elif node.data < value:\n",
    "            return self.find(value, node.right_child)\n",
    "        return node\n",
    "\n",
    "    def _delete_node(self, node):\n",
    "        parent_node = node.parent\n",
    "        num_child = self._num_children(node)\n",
    "\n",
    "        if num_child == 0:\n",
    "            # If there is no parent then 'node' is the root node.\n",
    "            # 'node' has no children, so set root to point to None.\n",
    "            if not parent_node:\n",
    "                self.root = None\n",
    "            elif parent_node.left_child == node:\n",
    "                parent_node.left_child = None\n",
    "            else:\n",
    "                parent_node.right_child = None\n",
    "            return parent_node\n",
    "\n",
    "        elif num_child == 1:\n",
    "            if node.left_child:\n",
    "                child = node.left_child\n",
    "            else:\n",
    "                child = node.right_child\n",
    "\n",
    "            if not parent_node:\n",
    "                self.root = child\n",
    "            elif parent_node.left_child == node:\n",
    "                parent_node.left_child = child\n",
    "            else:\n",
    "                parent_node.right_child = child\n",
    "            child.parent = parent_node\n",
    "            return parent_node\n",
    "\n",
    "        else:\n",
    "            successor = self._max_node(node.left_child)\n",
    "            node.data = successor.data\n",
    "            self._delete_node(successor)\n",
    "\n",
    "    @staticmethod\n",
    "    def _num_children(node):\n",
    "        num_children = 0\n",
    "        if node.left_child:\n",
    "            num_children += 1\n",
    "        if node.right_child:\n",
    "            num_children += 1\n",
    "        return num_children\n",
    "\n",
    "    def _walk_up(self, node):\n",
    "        if not node:\n",
    "            return\n",
    "        else:\n",
    "            self._check_node(node)\n",
    "            return self._walk_up(node.parent)\n",
    "\n",
    "    def _check_node(self, node):\n",
    "        left_height = -1\n",
    "        right_height = -1\n",
    "        if node.left_child:\n",
    "            left_height = node.left_child.height\n",
    "        if node.right_child:\n",
    "            right_height = node.right_child.height\n",
    "        if abs(left_height - right_height) > 1:\n",
    "            if left_height < right_height:\n",
    "                self.left_rotate(node, node.right_child)\n",
    "            else:\n",
    "                self.right_rotate(node, node.left_child)\n",
    "        else:\n",
    "            node.height = max(left_height, right_height) + 1\n",
    "\n",
    "    def left_rotate(self, node, child_node):\n",
    "        if child_node.left_child:\n",
    "            node.right_child = child_node.left_child\n",
    "            node.right_child.parent = node\n",
    "        else:\n",
    "            node.right_child = None\n",
    "\n",
    "        if node != self.root:\n",
    "            child_node.parent = node.parent\n",
    "            if node.parent.right_child == node:\n",
    "                node.parent.right_child = child_node\n",
    "            else:\n",
    "                node.parent.left_child = child_node\n",
    "        else:\n",
    "            child_node.parent = None\n",
    "            # because we are not replacing the parent node('node') with the\n",
    "            # child node('child_node'), self.root is still pointing at the parent node.\n",
    "            self.root = child_node\n",
    "\n",
    "        child_node.left_child = node\n",
    "        node.parent = child_node\n",
    "\n",
    "        node.height -= 1\n",
    "\n",
    "    def right_rotate(self, node, child_node):\n",
    "        if child_node.right_child:\n",
    "            node.left_child = child_node.right_child\n",
    "            node.left_child.parent = node\n",
    "        else:\n",
    "            node.left_child = None\n",
    "\n",
    "        if node != self.root:\n",
    "            child_node.parent = node.parent\n",
    "            if node.parent.right_child == node:\n",
    "                node.parent.right_child = child_node\n",
    "            else:\n",
    "                node.parent.left_child = child_node\n",
    "            # node.parent.left_child = child_node\n",
    "        else:\n",
    "            child_node.parent = None\n",
    "            self.root = child_node\n",
    "\n",
    "        child_node.right_child = node\n",
    "        node.parent = child_node\n",
    "\n",
    "        node.height -= 1\n",
    "\n",
    "    def _min_node(self, node):\n",
    "        if node.left_child:\n",
    "            return self._min_node(node.left_child)\n",
    "        return node\n",
    "\n",
    "    def _max_node(self, node):\n",
    "        if node.right_child:\n",
    "            return self._max_node(node.right_child)\n",
    "        return node\n",
    "\n",
    "    def traverse(self, method='in'):\n",
    "        if not self.root:\n",
    "            return iter(())\n",
    "        if method == 'in':\n",
    "            return self._traverse_inorder(self.root)\n",
    "        elif method == 'pre':\n",
    "            return self._traverse_preorder(self.root)\n",
    "        elif method == 'post':\n",
    "            return self._traverse_postorder(self.root)\n",
    "        else:\n",
    "            raise ValueError('method must be either \"in\", \"pre\" or \"post\".')\n",
    "\n",
    "    # left subtree -> root -> right subtree\n",
    "    def _traverse_inorder(self, node):\n",
    "        if node.left_child:\n",
    "            yield from self._traverse_inorder(node.left_child)\n",
    "        yield node\n",
    "        if node.right_child:\n",
    "            yield from self._traverse_inorder(node.right_child)\n",
    "\n",
    "    # root -> left subtree -> right subtree\n",
    "    def _traverse_preorder(self, node):\n",
    "        yield node\n",
    "        if node.left_child:\n",
    "            yield from self._traverse_inorder(node.left_child)\n",
    "        if node.right_child:\n",
    "            yield from self._traverse_inorder(node.right_child)\n",
    "\n",
    "    # left subtree -> right subtree -> root\n",
    "    def _traverse_postorder(self, node):\n",
    "        if node.left_child:\n",
    "            yield from self._traverse_inorder(node.left_child)\n",
    "        if node.right_child:\n",
    "            yield from self._traverse_inorder(node.right_child)\n",
    "        yield node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_010685819.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>731349093</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3555</td>\n",
       "      <td>Beta vulgaris subsp. vulgaris</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MDSTAGGKAKKGAGGRKGGGPKKKPVSRSVKAGLQFPVGRIGRYLK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_563627.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18378832</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MARTKHRVTRSQPRNQTDAAGASSSQAAGPTTTPTRRGGEGGDNTQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_001190852.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>334186954</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MAGRGKQLGSGAAKKSTSRSSKAGLQFPVGRIARFLKAGKYAERVG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NP_175517.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15223708</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MAGRGKTLGSGSAKKATTRSSKAGLQFPVGRIARFLKKGKYAERVG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP_188703.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15232330</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MAGRGKTLGSGVAKKSTSRSSKAGLQFPVGRIARFLKNGKYATRVG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession type variant_group variant  doublet         gi  \\\n",
       "0  XP_010685819.1  H2A          cH2A    cH2A      NaN  731349093   \n",
       "1     NP_563627.1   H3         cenH3   cenH3      NaN   18378832   \n",
       "2  NP_001190852.1  H2A          cH2A    cH2A      NaN  334186954   \n",
       "3     NP_175517.1  H2A          cH2A    cH2A      NaN   15223708   \n",
       "4     NP_188703.1  H2A          cH2A    cH2A      NaN   15232330   \n",
       "\n",
       "   ncbi_gene_id hgnc_gene_name  taxonomy_id                       organism  \\\n",
       "0           NaN            NaN         3555  Beta vulgaris subsp. vulgaris   \n",
       "1           NaN            NaN         3702           Arabidopsis thaliana   \n",
       "2           NaN            NaN         3702           Arabidopsis thaliana   \n",
       "3           NaN            NaN         3702           Arabidopsis thaliana   \n",
       "4           NaN            NaN         3702           Arabidopsis thaliana   \n",
       "\n",
       "         phylum          class taxonomy_group info references  \\\n",
       "0  Streptophyta  Magnoliopsida            NaN  NaN        NaN   \n",
       "1  Streptophyta  Magnoliopsida            NaN  NaN        NaN   \n",
       "2  Streptophyta  Magnoliopsida            NaN  NaN        NaN   \n",
       "3  Streptophyta  Magnoliopsida            NaN  NaN        NaN   \n",
       "4  Streptophyta  Magnoliopsida            NaN  NaN        NaN   \n",
       "\n",
       "                                            sequence  \n",
       "0  MDSTAGGKAKKGAGGRKGGGPKKKPVSRSVKAGLQFPVGRIGRYLK...  \n",
       "1  MARTKHRVTRSQPRNQTDAAGASSSQAAGPTTTPTRRGGEGGDNTQ...  \n",
       "2  MAGRGKQLGSGAAKKSTSRSSKAGLQFPVGRIARFLKAGKYAERVG...  \n",
       "3  MAGRGKTLGSGSAKKATTRSSKAGLQFPVGRIARFLKKGKYAERVG...  \n",
       "4  MAGRGKTLGSGVAKKSTSRSSKAGLQFPVGRIARFLKNGKYATRVG...  "
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "his_table = pd.read_csv('histones.csv')\n",
    "his_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Arabidopsis thaliana\n"
     ]
    }
   ],
   "source": [
    "def get_species(acc, df):\n",
    "    request = df.loc[df['accession'] == acc]['organism']\n",
    "    try:\n",
    "        species = request.iloc[0]\n",
    "    except IndexError:\n",
    "        print('error, skip', acc, request)\n",
    "        return False\n",
    "    else:\n",
    "        return species\n",
    "\n",
    "a = get_species('NP_188703.1', his_table)\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cH2A\n"
     ]
    }
   ],
   "source": [
    "def get_variant(acc, df):\n",
    "    request = df.loc[df['accession'] == acc]['variant']\n",
    "    try:\n",
    "        variant = request.iloc[0]\n",
    "    except IndexError:\n",
    "        print('error, skip', acc, request)\n",
    "        return False\n",
    "    else:\n",
    "        return variant\n",
    "\n",
    "\n",
    "a = get_variant('NP_188703.1', his_table)\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error, skip sp|O17536.3|H14_CAEEL Series([], Name: organism, dtype: object)\n",
      "error, skip sp|O17536.3|H14_CAEEL Series([], Name: variant, dtype: object)\n",
      "error, skip sp|P23444.2|H1_MAIZE Series([], Name: organism, dtype: object)\n",
      "error, skip sp|P23444.2|H1_MAIZE Series([], Name: variant, dtype: object)\n",
      "error, skip sp|P02272.2|H2AV_CHICK Series([], Name: organism, dtype: object)\n",
      "error, skip sp|P02272.2|H2AV_CHICK Series([], Name: variant, dtype: object)\n",
      "error, skip sp|Q9LZ45.3|H2B9_ARATH Series([], Name: organism, dtype: object)\n",
      "error, skip sp|Q9LZ45.3|H2B9_ARATH Series([], Name: variant, dtype: object)\n",
      "error, skip sp|Q9FFC0.3|H2B10_ARATH Series([], Name: organism, dtype: object)\n",
      "error, skip sp|Q9FFC0.3|H2B10_ARATH Series([], Name: variant, dtype: object)\n",
      "error, skip sp|Q9LQQ4.3|H2B1_ARATH Series([], Name: organism, dtype: object)\n",
      "error, skip sp|Q9LQQ4.3|H2B1_ARATH Series([], Name: variant, dtype: object)\n",
      "error, skip sp|Q9LZT0.3|H2B7_ARATH Series([], Name: organism, dtype: object)\n",
      "error, skip sp|Q9LZT0.3|H2B7_ARATH Series([], Name: variant, dtype: object)\n",
      "error, skip sp|P02291.2|H2BS2_PARAN Series([], Name: organism, dtype: object)\n",
      "error, skip sp|P02291.2|H2BS2_PARAN Series([], Name: variant, dtype: object)\n",
      "error, skip sp|Q27750.3|H2BS2_PSAMI Series([], Name: organism, dtype: object)\n",
      "error, skip sp|Q27750.3|H2BS2_PSAMI Series([], Name: variant, dtype: object)\n",
      "error, skip sp|Q27749.3|H2BS1_PSAMI Series([], Name: organism, dtype: object)\n",
      "error, skip sp|Q27749.3|H2BS1_PSAMI Series([], Name: variant, dtype: object)\n",
      "error, skip sp|P10651.2|H33_SCHPO Series([], Name: organism, dtype: object)\n",
      "error, skip sp|P10651.2|H33_SCHPO Series([], Name: variant, dtype: object)\n",
      "error, skip sp|Q7XYZ0.3|H3_GRIJA Series([], Name: organism, dtype: object)\n",
      "error, skip sp|Q7XYZ0.3|H3_GRIJA Series([], Name: variant, dtype: object)\n",
      "0.5065290927886963\n",
      "304\n",
      "{'H1': 235, 'H2A': 601, 'H2B': 272, 'H3': 288, 'H4': 134}\n"
     ]
    }
   ],
   "source": [
    "# open files aligned by its type\n",
    "\n",
    "tree = AVLTree()\n",
    "\n",
    "path = './new_alignments'\n",
    "h1 = open(path + '/H1.fasta')\n",
    "h2a = open(path + '/H2A.fasta')\n",
    "h2b = open(path + '/H2B.fasta')\n",
    "h3 = open(path + '/H3.fasta')\n",
    "h4 = open(path + '/H4.fasta')\n",
    "\n",
    "\n",
    "# add histone alignments from files into the avl_tree. Each H1 histone is assigned with id that gives\n",
    "#remainder of 0 when devided by 5, H2A histone give remainders of 1, H2B - 2, H3 - 3, H4 - 4.\n",
    "#After obtaining following id increase by 5\n",
    "\n",
    "start = t.time()\n",
    "\n",
    "files = [h1, h2a, h2b, h3, h4]\n",
    "types = ['H1', 'H2A', 'H2B', 'H3', 'H4']\n",
    "rests = {'H1': 0, 'H2A': 1, 'H2B': 2, 'H3':  3, 'H4': 4}\n",
    "id_counter = {'H1': 0, 'H2A': 1, 'H2B': 2, 'H3':  3, 'H4': 4}\n",
    "his_counter = 0\n",
    "i = -1\n",
    "\n",
    "\n",
    "for file in files:\n",
    "    i += 1\n",
    "    cur_his_type = types[i]\n",
    "    for record in SeqIO.parse(file, 'fasta'):\n",
    "        species = get_species(record.id, his_table)\n",
    "        variant = get_variant(record.id, his_table)\n",
    "        if species and variant:\n",
    "            res_ind = [i for i in range(len(record.seq)) if record.seq[i] != '-']\n",
    "            his_id = id_counter[cur_his_type]\n",
    "            id_counter[cur_his_type] += 5\n",
    "            his_counter += 1\n",
    "            tree.insert(his_id, record, res_ind, species, variant)\n",
    "        else:\n",
    "            continue\n",
    "        \n",
    "                \n",
    "print(t.time() - start)\n",
    "print(his_counter)\n",
    "print(id_counter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>A_segid</th>\n",
       "      <th>A_resid</th>\n",
       "      <th>A_resname</th>\n",
       "      <th>B_segid</th>\n",
       "      <th>B_resid</th>\n",
       "      <th>B_resname</th>\n",
       "      <th>num_int</th>\n",
       "      <th>A_entity</th>\n",
       "      <th>B_entity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>202</td>\n",
       "      <td>A</td>\n",
       "      <td>44</td>\n",
       "      <td>G</td>\n",
       "      <td>B</td>\n",
       "      <td>44</td>\n",
       "      <td>K</td>\n",
       "      <td>2</td>\n",
       "      <td>H3</td>\n",
       "      <td>H4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>227</td>\n",
       "      <td>A</td>\n",
       "      <td>47</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>39</td>\n",
       "      <td>R</td>\n",
       "      <td>5</td>\n",
       "      <td>H3</td>\n",
       "      <td>H4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>228</td>\n",
       "      <td>A</td>\n",
       "      <td>47</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>44</td>\n",
       "      <td>K</td>\n",
       "      <td>1</td>\n",
       "      <td>H3</td>\n",
       "      <td>H4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>238</td>\n",
       "      <td>A</td>\n",
       "      <td>48</td>\n",
       "      <td>L</td>\n",
       "      <td>B</td>\n",
       "      <td>44</td>\n",
       "      <td>K</td>\n",
       "      <td>2</td>\n",
       "      <td>H3</td>\n",
       "      <td>H4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>239</td>\n",
       "      <td>A</td>\n",
       "      <td>48</td>\n",
       "      <td>L</td>\n",
       "      <td>G</td>\n",
       "      <td>115</td>\n",
       "      <td>L</td>\n",
       "      <td>2</td>\n",
       "      <td>H3</td>\n",
       "      <td>H2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8294</th>\n",
       "      <td>8294</td>\n",
       "      <td>H</td>\n",
       "      <td>118</td>\n",
       "      <td>Y</td>\n",
       "      <td>G</td>\n",
       "      <td>20</td>\n",
       "      <td>R</td>\n",
       "      <td>2</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8295</th>\n",
       "      <td>8295</td>\n",
       "      <td>H</td>\n",
       "      <td>118</td>\n",
       "      <td>Y</td>\n",
       "      <td>G</td>\n",
       "      <td>49</td>\n",
       "      <td>V</td>\n",
       "      <td>2</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8317</th>\n",
       "      <td>8317</td>\n",
       "      <td>H</td>\n",
       "      <td>121</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>20</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8324</th>\n",
       "      <td>8324</td>\n",
       "      <td>H</td>\n",
       "      <td>122</td>\n",
       "      <td>K</td>\n",
       "      <td>G</td>\n",
       "      <td>6</td>\n",
       "      <td>Q</td>\n",
       "      <td>2</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8325</th>\n",
       "      <td>8325</td>\n",
       "      <td>H</td>\n",
       "      <td>122</td>\n",
       "      <td>K</td>\n",
       "      <td>G</td>\n",
       "      <td>20</td>\n",
       "      <td>R</td>\n",
       "      <td>11</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1136 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0 A_segid  A_resid A_resname B_segid  B_resid B_resname  \\\n",
       "202          202       A       44         G       B       44         K   \n",
       "227          227       A       47         A       B       39         R   \n",
       "228          228       A       47         A       B       44         K   \n",
       "238          238       A       48         L       B       44         K   \n",
       "239          239       A       48         L       G      115         L   \n",
       "...          ...     ...      ...       ...     ...      ...       ...   \n",
       "8294        8294       H      118         Y       G       20         R   \n",
       "8295        8295       H      118         Y       G       49         V   \n",
       "8317        8317       H      121         A       G       20         R   \n",
       "8324        8324       H      122         K       G        6         Q   \n",
       "8325        8325       H      122         K       G       20         R   \n",
       "\n",
       "      num_int A_entity B_entity  \n",
       "202         2       H3       H4  \n",
       "227         5       H3       H4  \n",
       "228         1       H3       H4  \n",
       "238         2       H3       H4  \n",
       "239         2       H3      H2A  \n",
       "...       ...      ...      ...  \n",
       "8294        2      H2B      H2A  \n",
       "8295        2      H2B      H2A  \n",
       "8317        1      H2B      H2A  \n",
       "8324        2      H2B      H2A  \n",
       "8325       11      H2B      H2A  \n",
       "\n",
       "[1136 rows x 10 columns]"
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read contact table and delit letters p and d from histone names\n",
    "\n",
    "conts = pd.read_csv('../histone_contacts.csv')\n",
    "conts = conts[(conts['A_entity'] != conts['B_entity'])]\n",
    "\n",
    "conts['A_entity'] = [his[:-1] for his in conts['A_entity']]\n",
    "conts['B_entity'] = [his[:-1] for his in conts['B_entity']]\n",
    "conts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.7746615409851074\n"
     ]
    }
   ],
   "source": [
    "# search for synchronous replacement of contacting residues in the aligned histone sequences\n",
    "\n",
    "start = t.time()\n",
    "\n",
    "\n",
    "res_dict = {key: [] for key in ['species', 'a_resid', 'b_resid', 'a_entity', 'b_entity', 'a_resname',\\\n",
    "                        'b_resname', 'a_new_resname', 'b_new_resname', 'a_variant', 'b_variant',\\\n",
    "                               'a_accsession', 'b_accsession']}\n",
    "\n",
    "i = -1\n",
    "\n",
    "for j, row in conts.iterrows():\n",
    "    a_his, b_his = row[8], row[9]\n",
    "    a_resi, b_resi = row[2] - 1, row[5] - 1\n",
    "    a_resid, b_resid = row['A_resid'], row['B_resid']\n",
    "    a_resname, b_resname = row[3], row[6]\n",
    "    a_start, b_start = rests[a_his], rests[b_his]\n",
    "    a_stop, b_stop = id_counter[a_his], id_counter[b_his]\n",
    "    i += 1\n",
    "\n",
    "    while a_start < a_stop:\n",
    "        a_node = tree.find(a_start, tree.root)\n",
    "        a_start += 5\n",
    "        a_seq = a_node.value.seq\n",
    "        try:\n",
    "            a_new_resname = a_seq[a_node.res_ind[a_resi]]\n",
    "        except IndexError:\n",
    "            continue\n",
    "        else:\n",
    "            if a_new_resname != a_resname:\n",
    "                \n",
    "                while b_start < b_stop:\n",
    "                    b_node = tree.find(b_start, tree.root)\n",
    "                    b_start += 5\n",
    "                    b_seq = b_node.value.seq\n",
    "                    try:\n",
    "                        b_new_resname = b_seq[b_node.res_ind[b_resi]]\n",
    "                    except IndexError:\n",
    "                        continue\n",
    "                    else:\n",
    "                        if b_new_resname != a_new_resname and a_node.species == b_node.species:\n",
    "                            res_tup = (a_node.species, a_resid, b_resid, a_his, b_his,\n",
    "                                                 a_resname, b_resname, a_new_resname, b_new_resname,\n",
    "                                       a_node.variant, b_node.variant, a_node.value.id, b_node.value.id)\n",
    "                            \n",
    "                            for item, key in zip(res_tup, res_dict):\n",
    "                                res_dict[key].append(item)\n",
    "                                \n",
    "        \n",
    "print(t.time() - start)\n",
    "\n",
    "# scrypt returns species, contacting histones in 1KX5, position and residue of each histone in 1KX5, \n",
    "# new residues and replaced hisone variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>species</th>\n",
       "      <th>a_resid</th>\n",
       "      <th>b_resid</th>\n",
       "      <th>a_entity</th>\n",
       "      <th>b_entity</th>\n",
       "      <th>a_resname</th>\n",
       "      <th>b_resname</th>\n",
       "      <th>a_new_resname</th>\n",
       "      <th>b_new_resname</th>\n",
       "      <th>a_variant</th>\n",
       "      <th>b_variant</th>\n",
       "      <th>a_accsession</th>\n",
       "      <th>b_accsession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>H3</td>\n",
       "      <td>H4</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>T</td>\n",
       "      <td>V</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cH4</td>\n",
       "      <td>NP_523730.2</td>\n",
       "      <td>NP_001027352.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>47</td>\n",
       "      <td>39</td>\n",
       "      <td>H3</td>\n",
       "      <td>H4</td>\n",
       "      <td>A</td>\n",
       "      <td>R</td>\n",
       "      <td>Q</td>\n",
       "      <td>A</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cH4</td>\n",
       "      <td>NP_523730.2</td>\n",
       "      <td>NP_001027352.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>47</td>\n",
       "      <td>44</td>\n",
       "      <td>H3</td>\n",
       "      <td>H4</td>\n",
       "      <td>A</td>\n",
       "      <td>K</td>\n",
       "      <td>Q</td>\n",
       "      <td>V</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cH4</td>\n",
       "      <td>NP_523730.2</td>\n",
       "      <td>NP_001027352.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>48</td>\n",
       "      <td>44</td>\n",
       "      <td>H3</td>\n",
       "      <td>H4</td>\n",
       "      <td>L</td>\n",
       "      <td>K</td>\n",
       "      <td>T</td>\n",
       "      <td>V</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cH4</td>\n",
       "      <td>NP_563627.1</td>\n",
       "      <td>NP_180441.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>48</td>\n",
       "      <td>115</td>\n",
       "      <td>H3</td>\n",
       "      <td>H2A</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>T</td>\n",
       "      <td>P</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>H2A.Z</td>\n",
       "      <td>NP_563627.1</td>\n",
       "      <td>NP_193093.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1628</th>\n",
       "      <td>1628</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>122</td>\n",
       "      <td>6</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2A</td>\n",
       "      <td>K</td>\n",
       "      <td>Q</td>\n",
       "      <td>M</td>\n",
       "      <td>H</td>\n",
       "      <td>H2B.W</td>\n",
       "      <td>H2A.L</td>\n",
       "      <td>DAA13058.1</td>\n",
       "      <td>NP_001071426.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1629</th>\n",
       "      <td>1629</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>122</td>\n",
       "      <td>20</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2A</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>M</td>\n",
       "      <td>S</td>\n",
       "      <td>H2B.W</td>\n",
       "      <td>H2A.X</td>\n",
       "      <td>DAA13058.1</td>\n",
       "      <td>NP_001073248.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1630</th>\n",
       "      <td>1630</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>122</td>\n",
       "      <td>20</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2A</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>M</td>\n",
       "      <td>S</td>\n",
       "      <td>H2B.W</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>DAA13058.1</td>\n",
       "      <td>NP_001192525.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1631</th>\n",
       "      <td>1631</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>122</td>\n",
       "      <td>20</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2A</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>M</td>\n",
       "      <td>S</td>\n",
       "      <td>H2B.W</td>\n",
       "      <td>H2A.B</td>\n",
       "      <td>DAA13058.1</td>\n",
       "      <td>NP_001069373.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1632</th>\n",
       "      <td>1632</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>122</td>\n",
       "      <td>20</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2A</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>M</td>\n",
       "      <td>R</td>\n",
       "      <td>H2B.W</td>\n",
       "      <td>H2A.L</td>\n",
       "      <td>DAA13058.1</td>\n",
       "      <td>NP_001071426.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1633 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index                  species  a_resid  b_resid a_entity b_entity  \\\n",
       "0         0  Drosophila melanogaster       44       44       H3       H4   \n",
       "1         1  Drosophila melanogaster       47       39       H3       H4   \n",
       "2         2  Drosophila melanogaster       47       44       H3       H4   \n",
       "3         3     Arabidopsis thaliana       48       44       H3       H4   \n",
       "4         4     Arabidopsis thaliana       48      115       H3      H2A   \n",
       "...     ...                      ...      ...      ...      ...      ...   \n",
       "1628   1628               Bos taurus      122        6      H2B      H2A   \n",
       "1629   1629               Bos taurus      122       20      H2B      H2A   \n",
       "1630   1630               Bos taurus      122       20      H2B      H2A   \n",
       "1631   1631               Bos taurus      122       20      H2B      H2A   \n",
       "1632   1632               Bos taurus      122       20      H2B      H2A   \n",
       "\n",
       "     a_resname b_resname a_new_resname b_new_resname a_variant b_variant  \\\n",
       "0            G         K             T             V     cenH3       cH4   \n",
       "1            A         R             Q             A     cenH3       cH4   \n",
       "2            A         K             Q             V     cenH3       cH4   \n",
       "3            L         K             T             V     cenH3       cH4   \n",
       "4            L         L             T             P     cenH3     H2A.Z   \n",
       "...        ...       ...           ...           ...       ...       ...   \n",
       "1628         K         Q             M             H     H2B.W     H2A.L   \n",
       "1629         K         R             M             S     H2B.W     H2A.X   \n",
       "1630         K         R             M             S     H2B.W      cH2A   \n",
       "1631         K         R             M             S     H2B.W     H2A.B   \n",
       "1632         K         R             M             R     H2B.W     H2A.L   \n",
       "\n",
       "     a_accsession    b_accsession  \n",
       "0     NP_523730.2  NP_001027352.1  \n",
       "1     NP_523730.2  NP_001027352.1  \n",
       "2     NP_523730.2  NP_001027352.1  \n",
       "3     NP_563627.1     NP_180441.1  \n",
       "4     NP_563627.1     NP_193093.1  \n",
       "...           ...             ...  \n",
       "1628   DAA13058.1  NP_001071426.1  \n",
       "1629   DAA13058.1  NP_001073248.1  \n",
       "1630   DAA13058.1  NP_001192525.1  \n",
       "1631   DAA13058.1  NP_001069373.1  \n",
       "1632   DAA13058.1  NP_001071426.1  \n",
       "\n",
       "[1633 rows x 14 columns]"
      ]
     },
     "execution_count": 262,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res = pd.DataFrame(res_dict).reset_index()\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1633 entries, 0 to 1632\n",
      "Data columns (total 14 columns):\n",
      " #   Column         Non-Null Count  Dtype \n",
      "---  ------         --------------  ----- \n",
      " 0   index          1633 non-null   int64 \n",
      " 1   species        1633 non-null   object\n",
      " 2   a_resid        1633 non-null   int64 \n",
      " 3   b_resid        1633 non-null   int64 \n",
      " 4   a_entity       1633 non-null   object\n",
      " 5   b_entity       1633 non-null   object\n",
      " 6   a_resname      1633 non-null   object\n",
      " 7   b_resname      1633 non-null   object\n",
      " 8   a_new_resname  1633 non-null   object\n",
      " 9   b_new_resname  1633 non-null   object\n",
      " 10  a_variant      1633 non-null   object\n",
      " 11  b_variant      1633 non-null   object\n",
      " 12  a_accsession   1633 non-null   object\n",
      " 13  b_accsession   1633 non-null   object\n",
      "dtypes: int64(3), object(11)\n",
      "memory usage: 178.7+ KB\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['Drosophila melanogaster', 'Arabidopsis thaliana',\n",
       "       'Trypanosoma brucei brucei TREU927',\n",
       "       'Tetrahymena thermophila SB210', 'Giardia intestinalis',\n",
       "       'Bos taurus', 'Homo sapiens'], dtype=object)"
      ]
     },
     "execution_count": 263,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.info()\n",
    "res['species'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:moldyn]",
   "language": "python",
   "name": "conda-env-moldyn-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
