{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Phylo\n",
    "import pandas as pd\n",
    "from Bio.Phylo.TreeConstruction import DistanceCalculator \n",
    "from Bio import AlignIO\n",
    "from Bio.Phylo.TreeConstruction import DistanceTreeConstructor\n",
    "from ete3 import PhyloTree\n",
    "from ete3 import Tree\n",
    "from ete3 import NCBITaxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi = NCBITaxa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('new_alignments/H1.fasta') as h1_file, open('new_alignments/H2A.fasta') as h2a_file,\\\n",
    "open('new_alignments/H2B.fasta') as h2b_file, open('new_alignments/H3.fasta') as h3_file,\\\n",
    "open('new_alignments/H4.fasta') as h4_file:\n",
    "\n",
    "    h1_al = AlignIO.read(h1_file, 'fasta')\n",
    "    h2a_al = AlignIO.read(h2a_file, 'fasta')\n",
    "    h2b_al = AlignIO.read(h2b_file, 'fasta')\n",
    "    h3_al = AlignIO.read(h3_file, 'fasta')\n",
    "    h4_al = AlignIO.read(h4_file, 'fasta')\n",
    " \n",
    "\n",
    "al_refs = [h1_al, h2a_al, h2b_al, h3_al, h4_al]\n",
    "names = ['H1.nwk', 'H2A.nwk', 'H2B.nwk', 'H3.nwk', 'H4.nwk']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warnning: [57] terminal nodes could not be found in the alignment.\n"
     ]
    }
   ],
   "source": [
    "for alignment, name in zip(al_refs, names):\n",
    "    dm = DistanceCalculator('identity').get_distance(alignment)\n",
    "    constructor = DistanceTreeConstructor()\n",
    "    nj_tree = constructor.nj(dm)\n",
    "    path = 'trees/' + name\n",
    "    Phylo.write(nj_tree, path, 'newick')\n",
    "    tree = PhyloTree(newick='trees/H1.nwk', alignment='new_alignments/H1.fasta', alg_format='fasta',\n",
    "                     format=3)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_010685819.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>731349093</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3555</td>\n",
       "      <td>Beta vulgaris subsp. vulgaris</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MDSTAGGKAKKGAGGRKGGGPKKKPVSRSVKAGLQFPVGRIGRYLK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_563627.1</td>\n",
       "      <td>H3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18378832</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MARTKHRVTRSQPRNQTDAAGASSSQAAGPTTTPTRRGGEGGDNTQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_001190852.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>334186954</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MAGRGKQLGSGAAKKSTSRSSKAGLQFPVGRIARFLKAGKYAERVG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NP_175517.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15223708</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MAGRGKTLGSGSAKKATTRSSKAGLQFPVGRIARFLKKGKYAERVG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP_188703.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15232330</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MAGRGKTLGSGVAKKSTSRSSKAGLQFPVGRIARFLKNGKYATRVG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession type variant_group variant  doublet         gi  \\\n",
       "0  XP_010685819.1  H2A          cH2A    cH2A      NaN  731349093   \n",
       "1     NP_563627.1   H3         cenH3   cenH3      NaN   18378832   \n",
       "2  NP_001190852.1  H2A          cH2A    cH2A      NaN  334186954   \n",
       "3     NP_175517.1  H2A          cH2A    cH2A      NaN   15223708   \n",
       "4     NP_188703.1  H2A          cH2A    cH2A      NaN   15232330   \n",
       "\n",
       "   ncbi_gene_id hgnc_gene_name  taxonomy_id                       organism  \\\n",
       "0           NaN            NaN         3555  Beta vulgaris subsp. vulgaris   \n",
       "1           NaN            NaN         3702           Arabidopsis thaliana   \n",
       "2           NaN            NaN         3702           Arabidopsis thaliana   \n",
       "3           NaN            NaN         3702           Arabidopsis thaliana   \n",
       "4           NaN            NaN         3702           Arabidopsis thaliana   \n",
       "\n",
       "         phylum          class taxonomy_group info references  \\\n",
       "0  Streptophyta  Magnoliopsida            NaN  NaN        NaN   \n",
       "1  Streptophyta  Magnoliopsida            NaN  NaN        NaN   \n",
       "2  Streptophyta  Magnoliopsida            NaN  NaN        NaN   \n",
       "3  Streptophyta  Magnoliopsida            NaN  NaN        NaN   \n",
       "4  Streptophyta  Magnoliopsida            NaN  NaN        NaN   \n",
       "\n",
       "                                            sequence  \n",
       "0  MDSTAGGKAKKGAGGRKGGGPKKKPVSRSVKAGLQFPVGRIGRYLK...  \n",
       "1  MARTKHRVTRSQPRNQTDAAGASSSQAAGPTTTPTRRGGEGGDNTQ...  \n",
       "2  MAGRGKQLGSGAAKKSTSRSSKAGLQFPVGRIARFLKAGKYAERVG...  \n",
       "3  MAGRGKTLGSGSAKKATTRSSKAGLQFPVGRIARFLKKGKYAERVG...  \n",
       "4  MAGRGKTLGSGVAKKSTSRSSKAGLQFPVGRIARFLKNGKYATRVG...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>species</th>\n",
       "      <th>a_resid</th>\n",
       "      <th>b_resid</th>\n",
       "      <th>a_entity</th>\n",
       "      <th>b_entity</th>\n",
       "      <th>a_resname</th>\n",
       "      <th>b_resname</th>\n",
       "      <th>a_new_resname</th>\n",
       "      <th>b_new_resname</th>\n",
       "      <th>a_variant</th>\n",
       "      <th>b_variant</th>\n",
       "      <th>a_accsession</th>\n",
       "      <th>b_accsession</th>\n",
       "      <th>a_column</th>\n",
       "      <th>b_column</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>105</td>\n",
       "      <td>Tetrahymena thermophila SB210</td>\n",
       "      <td>25</td>\n",
       "      <td>73</td>\n",
       "      <td>H4</td>\n",
       "      <td>H3</td>\n",
       "      <td>N</td>\n",
       "      <td>E</td>\n",
       "      <td>S</td>\n",
       "      <td>D</td>\n",
       "      <td>cH4</td>\n",
       "      <td>cH3</td>\n",
       "      <td>XP_001016593.1</td>\n",
       "      <td>XP_001016594.3</td>\n",
       "      <td>27</td>\n",
       "      <td>378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>106</td>\n",
       "      <td>Tetrahymena thermophila SB210</td>\n",
       "      <td>25</td>\n",
       "      <td>73</td>\n",
       "      <td>H4</td>\n",
       "      <td>H3</td>\n",
       "      <td>N</td>\n",
       "      <td>E</td>\n",
       "      <td>S</td>\n",
       "      <td>D</td>\n",
       "      <td>cH4</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>XP_001016593.1</td>\n",
       "      <td>XP_001008397.1</td>\n",
       "      <td>27</td>\n",
       "      <td>378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>109</td>\n",
       "      <td>Trypanosoma brucei brucei TREU927</td>\n",
       "      <td>41</td>\n",
       "      <td>105</td>\n",
       "      <td>H4</td>\n",
       "      <td>H3</td>\n",
       "      <td>G</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>cH4</td>\n",
       "      <td>cH3</td>\n",
       "      <td>XP_951561.1</td>\n",
       "      <td>XP_001218942.1</td>\n",
       "      <td>43</td>\n",
       "      <td>417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>110</td>\n",
       "      <td>Tetrahymena thermophila SB210</td>\n",
       "      <td>49</td>\n",
       "      <td>121</td>\n",
       "      <td>H4</td>\n",
       "      <td>H3</td>\n",
       "      <td>L</td>\n",
       "      <td>P</td>\n",
       "      <td>F</td>\n",
       "      <td>S</td>\n",
       "      <td>cH4</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>XP_001016593.1</td>\n",
       "      <td>XP_001011273.1</td>\n",
       "      <td>51</td>\n",
       "      <td>433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>111</td>\n",
       "      <td>Tetrahymena thermophila SB210</td>\n",
       "      <td>49</td>\n",
       "      <td>121</td>\n",
       "      <td>H4</td>\n",
       "      <td>H3</td>\n",
       "      <td>L</td>\n",
       "      <td>P</td>\n",
       "      <td>F</td>\n",
       "      <td>T</td>\n",
       "      <td>cH4</td>\n",
       "      <td>cH3</td>\n",
       "      <td>XP_001016593.1</td>\n",
       "      <td>XP_001016594.3</td>\n",
       "      <td>51</td>\n",
       "      <td>433</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                            species  a_resid  b_resid a_entity  \\\n",
       "0         105      Tetrahymena thermophila SB210       25       73       H4   \n",
       "1         106      Tetrahymena thermophila SB210       25       73       H4   \n",
       "2         109  Trypanosoma brucei brucei TREU927       41      105       H4   \n",
       "3         110      Tetrahymena thermophila SB210       49      121       H4   \n",
       "4         111      Tetrahymena thermophila SB210       49      121       H4   \n",
       "\n",
       "  b_entity a_resname b_resname a_new_resname b_new_resname a_variant  \\\n",
       "0       H3         N         E             S             D       cH4   \n",
       "1       H3         N         E             S             D       cH4   \n",
       "2       H3         G         E             A             A       cH4   \n",
       "3       H3         L         P             F             S       cH4   \n",
       "4       H3         L         P             F             T       cH4   \n",
       "\n",
       "  b_variant    a_accsession    b_accsession  a_column  b_column  \n",
       "0       cH3  XP_001016593.1  XP_001016594.3        27       378  \n",
       "1      H3.3  XP_001016593.1  XP_001008397.1        27       378  \n",
       "2       cH3     XP_951561.1  XP_001218942.1        43       417  \n",
       "3     cenH3  XP_001016593.1  XP_001011273.1        51       433  \n",
       "4       cH3  XP_001016593.1  XP_001016594.3        51       433  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "his_df, subs_df = pd.read_csv('histones.csv'), pd.read_csv('subs_df_cleaned.csv')\n",
    "display(his_df.head())\n",
    "subs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{9986, 9606, 9615, 10141, 8355, 10029, 9646, 13616, 9913, 9785, 7227, 559292, 10181, 9031, 9544, 312017, 185431, 9823, 6239, 508771, 10090, 423536, 3702}\n"
     ]
    }
   ],
   "source": [
    "tax_ids = set()\n",
    "\n",
    "for i, row in subs_df.iterrows():\n",
    "    a_acc, b_acc = row['a_accsession'], row['b_accsession']\n",
    "    a_tax_id = his_df.loc[his_df['accession']==a_acc]['taxonomy_id']\n",
    "    b_tax_id = his_df.loc[his_df['accession']==b_acc]['taxonomy_id']\n",
    "    tax_ids.add(int(a_tax_id)), tax_ids.add(int(b_tax_id))\n",
    " \n",
    "xenopus_id = 8355\n",
    "tax_ids.add(xenopus_id)\n",
    "print(tax_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi = NCBITaxa()\n",
    "\n",
    "tree = ncbi.get_topology(tax_ids)\n",
    "with open ('pictures_for_article/tax_tree', 'w') as file:\n",
    "    print(tree.get_ascii(attributes=[\"sci_name\", \"rank\"]), file=file)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>species</th>\n",
       "      <th>a_resid</th>\n",
       "      <th>b_resid</th>\n",
       "      <th>a_entity</th>\n",
       "      <th>b_entity</th>\n",
       "      <th>a_resname</th>\n",
       "      <th>b_resname</th>\n",
       "      <th>a_new_resname</th>\n",
       "      <th>b_new_resname</th>\n",
       "      <th>a_variant</th>\n",
       "      <th>b_variant</th>\n",
       "      <th>a_accsession</th>\n",
       "      <th>b_accsession</th>\n",
       "      <th>a_column</th>\n",
       "      <th>b_column</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>105</td>\n",
       "      <td>Tetrahymena thermophila SB210</td>\n",
       "      <td>25</td>\n",
       "      <td>73</td>\n",
       "      <td>H4</td>\n",
       "      <td>H3</td>\n",
       "      <td>N</td>\n",
       "      <td>E</td>\n",
       "      <td>S</td>\n",
       "      <td>D</td>\n",
       "      <td>cH4</td>\n",
       "      <td>cH3</td>\n",
       "      <td>XP_001016593.1</td>\n",
       "      <td>XP_001016594.3</td>\n",
       "      <td>27</td>\n",
       "      <td>378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>106</td>\n",
       "      <td>Tetrahymena thermophila SB210</td>\n",
       "      <td>25</td>\n",
       "      <td>73</td>\n",
       "      <td>H4</td>\n",
       "      <td>H3</td>\n",
       "      <td>N</td>\n",
       "      <td>E</td>\n",
       "      <td>S</td>\n",
       "      <td>D</td>\n",
       "      <td>cH4</td>\n",
       "      <td>H3.3</td>\n",
       "      <td>XP_001016593.1</td>\n",
       "      <td>XP_001008397.1</td>\n",
       "      <td>27</td>\n",
       "      <td>378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>109</td>\n",
       "      <td>Trypanosoma brucei brucei TREU927</td>\n",
       "      <td>41</td>\n",
       "      <td>105</td>\n",
       "      <td>H4</td>\n",
       "      <td>H3</td>\n",
       "      <td>G</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>cH4</td>\n",
       "      <td>cH3</td>\n",
       "      <td>XP_951561.1</td>\n",
       "      <td>XP_001218942.1</td>\n",
       "      <td>43</td>\n",
       "      <td>417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>110</td>\n",
       "      <td>Tetrahymena thermophila SB210</td>\n",
       "      <td>49</td>\n",
       "      <td>121</td>\n",
       "      <td>H4</td>\n",
       "      <td>H3</td>\n",
       "      <td>L</td>\n",
       "      <td>P</td>\n",
       "      <td>F</td>\n",
       "      <td>S</td>\n",
       "      <td>cH4</td>\n",
       "      <td>cenH3</td>\n",
       "      <td>XP_001016593.1</td>\n",
       "      <td>XP_001011273.1</td>\n",
       "      <td>51</td>\n",
       "      <td>433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>111</td>\n",
       "      <td>Tetrahymena thermophila SB210</td>\n",
       "      <td>49</td>\n",
       "      <td>121</td>\n",
       "      <td>H4</td>\n",
       "      <td>H3</td>\n",
       "      <td>L</td>\n",
       "      <td>P</td>\n",
       "      <td>F</td>\n",
       "      <td>T</td>\n",
       "      <td>cH4</td>\n",
       "      <td>cH3</td>\n",
       "      <td>XP_001016593.1</td>\n",
       "      <td>XP_001016594.3</td>\n",
       "      <td>51</td>\n",
       "      <td>433</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                            species  a_resid  b_resid a_entity  \\\n",
       "0         105      Tetrahymena thermophila SB210       25       73       H4   \n",
       "1         106      Tetrahymena thermophila SB210       25       73       H4   \n",
       "2         109  Trypanosoma brucei brucei TREU927       41      105       H4   \n",
       "3         110      Tetrahymena thermophila SB210       49      121       H4   \n",
       "4         111      Tetrahymena thermophila SB210       49      121       H4   \n",
       "\n",
       "  b_entity a_resname b_resname a_new_resname b_new_resname a_variant  \\\n",
       "0       H3         N         E             S             D       cH4   \n",
       "1       H3         N         E             S             D       cH4   \n",
       "2       H3         G         E             A             A       cH4   \n",
       "3       H3         L         P             F             S       cH4   \n",
       "4       H3         L         P             F             T       cH4   \n",
       "\n",
       "  b_variant    a_accsession    b_accsession  a_column  b_column  \n",
       "0       cH3  XP_001016593.1  XP_001016594.3        27       378  \n",
       "1      H3.3  XP_001016593.1  XP_001008397.1        27       378  \n",
       "2       cH3     XP_951561.1  XP_001218942.1        43       417  \n",
       "3     cenH3  XP_001016593.1  XP_001011273.1        51       433  \n",
       "4       cH3  XP_001016593.1  XP_001016594.3        51       433  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subs_df = pd.read_csv('subs_df_cleaned.csv')\n",
    "df = pd.read_csv('histones.csv')\n",
    "df.head()\n",
    "subs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "75"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df['organism'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "22"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(subs_df['species'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tetrahymena thermophila SB210 cH4 cH3 cenH3 H3.3\n",
      "Trypanosoma brucei brucei TREU927 cH2A cH4 H2A.Z cH3 cH2B\n",
      "Saccharomyces cerevisiae S288C cH4 H2A.X H3.3 cenH3 H2A.Z cH2B\n",
      "Homo sapiens cH4 H2A.Z cH3 cH2B\n",
      "Arabidopsis thaliana cH2A H2A.W H2A.X cenH3 H2A.Z cH2B\n",
      "Caenorhabditis elegans cH2A cenH3\n",
      "Drosophila melanogaster cH2A cenH3 H2A.Z cH2B\n",
      "Perkinsus marinus ATCC 50983 cH2A H2A.X cenH3 H2A.Z\n",
      "Mus musculus cH2A H2A.P H2A.1 H2A.B H2A.X H2A.L cenH3 subH2B\n",
      "Macaca mulatta cH2A H2B.W H2A.P H2A.B subH2B\n",
      "Canis lupus familiaris H2B.W H2A.P H2B.1 H2A.L\n",
      "Ailuropoda melanoleuca H2B.W H2A.L H2A.P cH2B H2A.B subH2B\n",
      "Loxodonta africana H2A.B subH2B\n",
      "Sus scrofa H2A.L H2A.P cH2B H2A.B subH2B\n",
      "Bos taurus H2B.W H2A.L cH2B H2A.B subH2B H2A.X\n",
      "Cricetulus griseus cH2A H2A.L H2A.P cH2B H2A.B subH2B\n",
      "Heterocephalus glaber H2A.P H2A.B subH2B H2A.L\n",
      "Oryctolagus cuniculus H2B.W H2A.L\n",
      "Cavia porcellus subH2B H2A.L\n",
      "Toxoplasma gondii ME49 cH2A H2B.Z H2A.X H2A.Z\n",
      "Monodelphis domestica cH2A H2A.X subH2B\n",
      "Gallus gallus cH2B macroH2A\n"
     ]
    }
   ],
   "source": [
    "sp_variants = {}\n",
    "variants = set()\n",
    "\n",
    "for sp in subs_df['species'].unique():\n",
    "    req_a = set(subs_df.loc[subs_df['species']==sp]['a_variant'].unique())\n",
    "    req_b = set(subs_df.loc[subs_df['species']==sp]['b_variant'].unique())\n",
    "    variants = variants.union(req_a.union(req_b))\n",
    "    unique_variants = str(req_a.union(req_b)).replace('{', '').replace('}', '').replace(',', '').replace(\"'\", '')\n",
    "    print(sp, unique_variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'H2A.1',\n",
       " 'H2A.B',\n",
       " 'H2A.L',\n",
       " 'H2A.P',\n",
       " 'H2A.W',\n",
       " 'H2A.X',\n",
       " 'H2A.Z',\n",
       " 'H2B.1',\n",
       " 'H2B.W',\n",
       " 'H2B.Z',\n",
       " 'H3.3',\n",
       " 'cH2A',\n",
       " 'cH2B',\n",
       " 'cH3',\n",
       " 'cH4',\n",
       " 'cenH3',\n",
       " 'macroH2A',\n",
       " 'subH2B'}"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:moldyn]",
   "language": "python",
   "name": "conda-env-moldyn-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
